python-package/lightgbm/sklearn.py

```diff
@@ -176,6 +176,170 @@ def __call__(self, preds, dataset):
+    eval_class_weight : list or None, optional (default=None)
+        Class weights of eval data.
+    eval_init_score : list of arrays or None, optional (default=None)
+        Init score of eval data.
+    eval_group : list of arrays or None, optional (default=None)
+        Group data of eval data.
+    eval_metric : string, callable, list or None, optional (default=None)
+        If string, it should be a built-in evaluation metric to use.
+        If callable, it should be a custom evaluation metric, see note below for more details.
+        If list, it can be a list of built-in metrics, a list of custom evaluation metrics, or a mix of both.
+        In either case, the ``metric`` from the model parameters will be evaluated and used as well.
+        Default: 'l2' for LGBMRegressor, 'logloss' for LGBMClassifier, 'ndcg' for LGBMRanker.
+    early_stopping_rounds : int or None, optional (default=None)
+        Activates early stopping. The model will train until the validation score stops improving.
+        Validation score needs to improve at least every ``early_stopping_rounds`` round(s)
+        to continue training.
+        Requires at least one validation data and one metric.
+        If there's more than one, will check all of them. But the training data is ignored anyway.
+        To check only the first metric, set the ``first_metric_only`` parameter to ``True``
+        in additional parameters ``**kwargs`` of the model constructor.
+    verbose : bool or int, optional (default=True)
+        Requires at least one evaluation data.
+        If True, the eval metric on the eval set is printed at each boosting stage.
+        If int, the eval metric on the eval set is printed at every ``verbose`` boosting stage.
+        The last boosting stage or the boosting stage found by using ``early_stopping_rounds`` is also printed.
+
+        .. rubric:: Example
+
+        With ``verbose`` = 4 and at least one item in ``eval_set``,
+        an evaluation metric is printed every 4 (instead of 1) boosting stages.
+
+    feature_name : list of strings or 'auto', optional (default='auto')
+        Feature names.
+        If 'auto' and data is pandas DataFrame, data columns names are used.
+    categorical_feature : list of strings or int, or 'auto', optional (default='auto')
+        Categorical features.
+        If list of int, interpreted as indices.
+        If list of strings, interpreted as feature names (need to specify ``feature_name`` as well).
+        If 'auto' and data is pandas DataFrame, pandas unordered categorical columns are used.
+        All values in categorical features should be less than int32 max value (2147483647).
+        Large values could be memory consuming. Consider using consecutive integers starting from zero.
+        All negative values in categorical features will be treated as missing values.
+        The output cannot be monotonically constrained with respect to a categorical feature.
```

Comment:
I believe default values for these three params (`verbose`, `feature_name`, `categorical_feature`) are not true for Dask. Refer to https://github.com/microsoft/LightGBM/pull/3908#discussion_r571343767
> As far as I can tell, "category" columns will only be automatically treated as categorical features if you set categorical_feature = 'auto', which is not the default, and of course you can force it to happen by setting categorical_featue to the column names / indices in the parameters.

Reply:
There isn't any special handling in the Dask interface for categorical features.

It's passed through directly to sklearn. So whatever is true for sklearn should be true for Dask. That comment you linked is about the behavior of the sklearn estimators.

Since `verbose`, `feature_name`, and `categorical_feature` aren't in the signatures of the Dask estimators, but just passed to the `sklearn` estimators through `**kwargs`, their defaults are identical to those from sklearn

Reply:
OK, got it!

Will `'auto'` work for dask DataFrame?

Reply:
I expect that it will, since the chunks of a Dask DataFrame are each pandas dataframes, and since we saw debugging #3861 that Dask DataFrame preserves the "category" type of the underlying dataframes.

However, I haven't actually tested that yet. I think it could be tested by removing `categorical_feature` from `params` in the Dask tests, like https://github.com/microsoft/LightGBM/blob/846b512d892426d34fac0749ccfd2d7a31eef877/tests/python_package_test/test_dask.py#L464 and instead setting it to `"categorical_feature": "auto"`. I can try that in a bit.

Reply:
interesting! When I tried setting `categorical_feature = "auto"` in params, I got this error:

> E           lightgbm.basic.LightGBMError: categorical_feature is not a number,
E           if you want to use a column name,
E           please add the prefix "name:" to the column name

But that error is coming from `lgb.LGBMClassifier.fit()`. `lgb.DaskLGBMClassifier.fit()` is ok with it

![image](https://user-images.githubusercontent.com/7608904/107417700-7c41c800-6adb-11eb-8647-4cab04270b4b.png)

When I switch to passing `categorical_feature="auto"` in `.fit()` (instead of in the constructor), all the tests pass!

So:

1. yes, `auto` seems to work for the Dask interface so this doc is accurate
2. I should probably change the tests (in a separate PR) to pass `categorical_feature` to `.fit()`, not in the model constructor. I can see that that's what the sklearn tests do: https://github.com/microsoft/LightGBM/blob/846b512d892426d34fac0749ccfd2d7a31eef877/tests/python_package_test/test_sklearn.py#L497-L508. Looking through the `sklearn` module, I see that `categorical_feature` is used in `fit()` everywhere.

Reply:
> and instead setting it to "categorical_feature": "auto". I can try that in a bit.

> I should probably change the tests (in a separate PR) to pass categorical_feature to .fit(), not in the model constructor

But `'auto'` is default value, so you can just omit `categorical_feature argument in tests at all. And yes, let's leave this for a separate PR please.

Reply:
good point! added https://github.com/microsoft/LightGBM/pull/3932